File: server/services/smartMatchService.js
/**
 * Smart Match Algorithm (Transparent 100-Point Rule-Based Engine)
 * Evaluates vendor compatibility against user event requirements.
 */

const calculateSmartMatchScore = (vendor, eventRequirements) => {
  const {
    eventType = 'Wedding',
    city = 'Patna',
    guestCount = 200,
    targetBudget = 70000
  } = eventRequirements;

  let totalScore = 0;
  const reasons = [];

  // 1. Location Match (Max 20 pts)
  if (vendor.city && vendor.city.toLowerCase() === city.toLowerCase()) {
    totalScore += 20;
    reasons.push(`Located directly in ${vendor.city}`);
  } else {
    totalScore += 10;
    reasons.push(`Regional vendor serving ${city}`);
  }

  // 2. Budget Compatibility (Max 25 pts)
  const vendorPrice = Number(vendor.starting_price) || 0;
  if (targetBudget > 0) {
    const ratio = vendorPrice / targetBudget;
    if (ratio <= 1.0) {
      totalScore += 25;
      reasons.push(`Within target budget allocation (₹${vendorPrice.toLocaleString('en-IN')})`);
    } else if (ratio <= 1.15) {
      totalScore += 18;
      reasons.push(`Slightly above target budget (+${Math.round((ratio - 1) * 100)}%)`);
    } else if (ratio <= 1.35) {
      totalScore += 10;
      reasons.push(`Moderately above target budget`);
    } else {
      totalScore += 5;
      reasons.push(`Premium pricing tier`);
    }
  } else {
    totalScore += 20;
  }

  // 3. Capacity Compatibility (Max 20 pts - primarily Venues)
  if (vendor.category === 'Venue') {
    const cap = Number(vendor.max_capacity) || 0;
    if (cap >= guestCount && cap <= guestCount * 2) {
      totalScore += 20;
      reasons.push(`Optimal capacity for ${guestCount} guests (Max: ${cap})`);
    } else if (cap >= guestCount) {
      totalScore += 16;
      reasons.push(`Large capacity facility (Max: ${cap} guests)`);
    } else {
      totalScore += 6;
      reasons.push(`Capacity (${cap}) below requested guest count (${guestCount})`);
    }
  } else {
    // Non-venue vendors get full capacity points if they handle the event scale
    totalScore += 20;
    reasons.push(`Equipped for ${guestCount}+ guest event scale`);
  }

  // 4. Event Type Match (Max 20 pts)
  const supported = (vendor.supported_events || '').toLowerCase();
  if (supported.includes(eventType.toLowerCase())) {
    totalScore += 20;
    reasons.push(`Specialized experience in ${eventType} celebrations`);
  } else {
    totalScore += 10;
    reasons.push(`General celebration and event service`);
  }

  // 5. Rating & Reputation (Max 15 pts)
  const rating = Number(vendor.rating) || 4.0;
  if (rating >= 4.8) {
    totalScore += 15;
    reasons.push(`Top-rated customer satisfaction (${rating} ★)`);
  } else if (rating >= 4.5) {
    totalScore += 12;
    reasons.push(`High customer rating (${rating} ★)`);
  } else {
    totalScore += 8;
    reasons.push(`Verified service standard (${rating} ★)`);
  }

  return {
    score: Math.min(100, Math.max(30, totalScore)),
    reasons
  };
};

module.exports = {
  calculateSmartMatchScore
};

File: server/services/smart_match_service.py
from datetime import datetime

def calculate_vendor_smart_match(vendor, event_params):
    """
    Transparent Rule-Based Smart Match Algorithm (Max 100 points).
    Evaluates:
    1. Location Match (20 pts)
    2. Budget Compatibility (25 pts)
    3. Capacity Compatibility (20 pts)
    4. Service Match (15 pts)
    5. Event Type Match (10 pts)
    6. Availability (10 pts)
    """
    score = 0
    reasons = []
    category = vendor.category
    city = event_params.get('city', '').strip().lower()
    event_type = event_params.get('event_type', '').strip().lower()
    guest_count = int(event_params.get('guest_count', 100))
    total_budget = float(event_params.get('total_budget', 100000))
    category_budget = float(event_params.get('category_budget', total_budget * 0.25))

    vendor_price = float(vendor.starting_price) if vendor.starting_price is not None else 0.0

    # 1. Location Match (20 pts)
    vendor_city = vendor.city.strip().lower()
    if vendor_city == city:
        score += 20
        reasons.append(f"Located directly in {vendor.city}")
    else:
        # Nearby Tier-2/3 Bihar regions
        score += 8
        reasons.append(f"Servicing from nearby region ({vendor.city})")

    # 2. Budget Compatibility (25 pts)
    if vendor_price <= category_budget:
        score += 25
        savings = category_budget - vendor_price
        if savings > 0:
            reasons.append(f"Within budget (Saves ₹{int(savings):,})")
        else:
            reasons.append("Exact fit for category budget")
    elif vendor_price <= category_budget * 1.2:
        score += 15
        reasons.append("Slightly above category budget (+<20%)")
    elif vendor_price <= category_budget * 1.5:
        score += 8
        reasons.append("Moderate budget stretch (+20-50%)")
    else:
        score += 0
        reasons.append("Exceeds standard category allocation")

    # 3. Capacity Compatibility (20 pts)
    if category == 'Venue' and vendor.venue_details:
        max_cap = vendor.venue_details.max_capacity
        if max_cap >= guest_count:
            score += 20
            reasons.append(f"Suitable venue capacity ({max_cap} max for {guest_count} guests)")
        elif max_cap >= guest_count * 0.8:
            score += 10
            reasons.append(f"Close capacity fit ({max_cap} vs {guest_count} guests)")
        else:
            reasons.append(f"Capacity below requested guest count ({max_cap} max)")
    elif category == 'Catering':
        score += 20
        reasons.append(f"Equipped for large gatherings ({guest_count}+ guests)")
    else:
        # Other service types
        score += 20
        reasons.append("Full team ready for event scale")

    # 4. Service Match (15 pts)
    required_services = [s.strip().lower() for s in event_params.get('required_services', [])]
    if not required_services or category.lower() in required_services:
        score += 15
        reasons.append(f"Matches required '{category}' service")
    else:
        score += 5

    # 5. Event Type Match (10 pts)
    supported_types = [t.strip().lower() for t in vendor.supported_event_types.split(',')] if vendor.supported_event_types else []
    if not event_type or event_type in supported_types:
        score += 10
        reasons.append(f"Experienced in {event_params.get('event_type', 'similar')} events")
    else:
        score += 4

    # 6. Availability & Reputation (10 pts)
    if vendor.is_verified:
        score += 10
        reasons.append("Verified vendor & available on date")
    else:
        score += 5

    final_score = min(100, max(0, score))
    return {
        'match_score': final_score,
        'reasons': reasons,
        'vendor_id': vendor.id,
        'category': vendor.category,
        'business_name': vendor.business_name,
        'starting_price': vendor_price,
        'rating': float(vendor.rating) if vendor.rating else 4.5
    }
